### Import Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
import seaborn
seaborn.set()

### Load data

In [4]:
df=pd.read_csv('datasets/Multiple linear regression.csv')

In [5]:
df.sample(5)

SAT  Rand 1,2,3   GPA
72  1975           1  3.60
60  1964           1  3.49
35  1849           2  3.31
39  1831           2  3.37
14  1773           2  3.12

In [7]:
df.describe()

SAT  Rand 1,2,3        GPA
count    84.000000   84.000000  84.000000
mean   1845.273810    2.059524   3.330238
std     104.530661    0.855192   0.271617
min    1634.000000    1.000000   2.400000
25%    1772.000000    1.000000   3.190000
50%    1846.000000    2.000000   3.380000
75%    1934.000000    3.000000   3.502500
max    2050.000000    3.000000   3.810000

## Multiple Regression

### Dependent and Independent Variables

In [44]:
x=df[['SAT','Rand 1,2,3']]
y=df['GPA']

### Regression Itself

In [45]:
reg=LinearRegression()

In [46]:
reg.fit(x,y)

LinearRegression()

In [47]:
reg.predict([[3333,3]])

c:\Users\Ghost\anaconda3\envs\class\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([5.78247797])

### Coeffiecient

In [48]:
reg.coef_

array([ 0.00165354, -0.00826982])

### Intercept

In [49]:
reg.intercept_

0.29603261264909486

### R-Square

In [50]:
reg.score(x,y)

0.40668119528142843

### Adjusted R-Squared

#### Formula
$R^2_{adj.}=1-(1-R^2)*\frac{n-1}{n-p-1}$

In [51]:
def adjusted_RSquared(x,y,reg):
    n=x.shape[0]
    p=x.shape[1]
    r2=reg.score(x,y)
    suffix=(n-1)/(n-p-1)
    preffix=(1-r2)
    
    return 1-preffix*suffix

In [52]:
print(adjusted_RSquared(x,y,reg))

0.39203134825134023


### Feature Selection as R-Adjusted was low

In [53]:
from sklearn.feature_selection import f_regression

In [54]:
f_reg=f_regression(x,y)
f_reg

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

In [55]:
p_values=f_reg[1]

In [56]:
x_features=['SAT','RAND 1,2,3']
for i in range(2):
    print(f'{x_features[i]} : {p_values.round(4)[i]}')

SAT : 0.0
RAND 1,2,3 : 0.6763


##### As 0.676>0.05 so Rand 1,2,3 is useless

### Creating a Summary Table

In [58]:
reg_summary=pd.DataFrame(data=x.columns.values,columns=['Features'])
reg_summary

Features
0         SAT
1  Rand 1,2,3

In [60]:
reg_summary['Coefficeints']=reg.coef_
reg_summary['p_values']=p_values.round(4)

In [61]:
reg_summary

Features  Coefficeints  p_values
0         SAT      0.001654    0.0000
1  Rand 1,2,3     -0.008270    0.6763